velox/functions/sparksql/aggregates/DecimalSumAggregate.h

```diff
@@ -0,0 +1,170 @@
+
+namespace facebook::velox::functions::aggregate::sparksql {
+
+/// @tparam TInputType The raw input data type.
+/// @tparam TSumType The type of sum in the output of partial aggregation or the
+/// final output type of final aggregation.
+template <typename TInputType, typename TSumType>
+class DecimalSumAggregate {
+ public:
+  using InputType = Row<TInputType>;
+
+  using IntermediateType =
+      Row</*sum*/ TSumType,
+          /*isEmpty*/ bool>;
+
+  using OutputType = TSumType;
+
+  // Spark's decimal sum doesn't have the concept of a null group, each group is
+  // initialized with an initial value, where sum = 0 and isEmpty = true.
+  // Therefore, to maintain consistency, we need to use the parameter
+  // nonNullGroup in writeIntermediateResult to output a null group as sum =
+  // 0, isEmpty = true. nonNullGroup is only available when default-null
+  // behavior is disabled.
+  static constexpr bool default_null_behavior_ = false;
+
+  static bool toIntermediate(
+      exec::out_type<Row<TSumType, bool>>& out,
+      exec::optional_arg_type<TInputType> in) {
+    if (in.has_value()) {
+      out.copy_from(std::make_tuple(static_cast<TSumType>(in.value()), false));
+      return true;
+    }
+    return false;
+  }
+
+  // This struct stores the sum of input values, overflow during accumulation,
+  // and a bool value isEmpty used to indicate whether all inputs are null. The
+  // initial value of sum is 0. We need to keep sum unchanged if the input is
+  // null, as sum function ignores null input. If the isEmpty is true, then it
+  // means there were no values to begin with or all the values were null, so
+  // the result will be null. If the isEmpty is false, then if sum is nullopt
+  // that means an overflow has happened, it returns null.
+  struct AccumulatorType {
+    std::optional<int128_t> sum_{0};
+    int64_t overflow_{0};
+    bool isEmpty_{true};
+
+    AccumulatorType() = delete;
+
+    explicit AccumulatorType(HashStringAllocator* /*allocator*/) {}
+
+    std::optional<int128_t> computeFinalResult() const {
+      if (!sum_.has_value()) {
+        return std::nullopt;
+      }
+      auto adjustedSum =
+          DecimalUtil::adjustSumForOverflow(sum_.value(), overflow_);
+      uint8_t maxPrecision = std::is_same_v<TSumType, int128_t>
+          ? LongDecimalType::kMaxPrecision
+          : ShortDecimalType::kMaxPrecision;
+      if (adjustedSum.has_value() &&
+          DecimalUtil::valueInPrecisionRange(adjustedSum, maxPrecision)) {
+        return adjustedSum;
+      } else {
+        // Found overflow during computing adjusted sum.
+        return std::nullopt;
+      }
+    }
+
+    bool addInput(
+        HashStringAllocator* /*allocator*/,
+        exec::optional_arg_type<TInputType> data) {
+      if (!data.has_value()) {
+        return false;
+      }
+      int128_t result;
+      overflow_ +=
+          DecimalUtil::addWithOverflow(result, data.value(), sum_.value());
```

Comment:
What is 'sum_' optional is not set?

Reply:
Looks like sum_ always have initial value 0 and is never set to std::nullopt in this method. On the other hand, the comment above says "If the isEmpty is false, then if sum is nullopt that means an overflow has happened". So shouldn't there be a check for overflow and set sum_ to nullopt here? Same question for the combine() method.

Reply:
Yes, sum will always have initial value 0 here. In the Spark implementation, only during the combine and writeFinalResult stages is there a check for overflow. If an overflow occurs during the addInput process, the overflow information is stored in the overflow variable. It is possible that due to the presence of negative values, it might revert to a non-overflow state; therefore, there is no need to check for overflow at this point. In combine, we have checked `sum.has_value()` to check overflow.